lib/internal/event_target.js

```diff
@@ -560,94 +470,45 @@ class EventTarget {
+  addEventListener(type, callback, options = kEmptyObject) {
     if (!isEventTarget(this))
       throw new ERR_INVALID_THIS('EventTarget');
     if (arguments.length < 2)
       throw new ERR_MISSING_ARGS('type', 'listener');
 
-    // We validateOptions before the validateListener check because the spec
-    // requires us to hit getters.
-    const {
-      once,
-      capture,
-      passive,
-      signal,
-      isNodeStyleListener,
-      weak,
-    } = validateEventListenerOptions(options);
-
-    validateAbortSignal(signal, 'options.signal');
+    const { capture, passive, once, signal } = flattenMoreOptions(options)
+    type = webidl.converters.DOMString(type);
 
-    if (!validateEventListener(listener)) {
+    if (!validateEventListener(callback)) {
       // The DOM silently allows passing undefined as a second argument
       // No error code for this since it is a Warning
       // eslint-disable-next-line no-restricted-syntax
-      const w = new Error(`addEventListener called with ${listener}` +
+      const w = new Error(`addEventListener called with ${callback}` +
                           ' which has no effect.');
       w.name = 'AddEventListenerArgumentTypeWarning';
       w.target = this;
       w.type = type;
       process.emitWarning(w);
       return;
     }
-    type = webidl.converters.DOMString(type);
-
-    if (signal) {
-      if (signal.aborted) {
-        return;
-      }
-      // TODO(benjamingr) make this weak somehow? ideally the signal would
-      // not prevent the event target from GC.
-      signal.addEventListener('abort', () => {
-        this.removeEventListener(type, listener, options);
-      }, { once: true, [kWeakHandler]: this });
-    }
-
-    let root = this[kEvents].get(type);
-
-    if (root === undefined) {
-      root = { size: 1, next: undefined };
-      // This is the first handler in our linked list.
-      new Listener(root, listener, once, capture, passive,
-                   isNodeStyleListener, weak);
-      this[kNewListener](
-        root.size,
-        type,
-        listener,
-        once,
-        capture,
-        passive,
-        weak);
-      this[kEvents].set(type, root);
-      return;
-    }
 
-    let handler = root.next;
-    let previous = root;
-
-    // We have to walk the linked list to see if we have a match
-    while (handler !== undefined && !handler.same(listener, capture)) {
-      previous = handler;
-      handler = handler.next;
-    }
-
-    if (handler !== undefined) { // Duplicate! Ignore
-      return;
-    }
-
-    new Listener(previous, listener, once, capture, passive,
-                 isNodeStyleListener, weak);
-    root.size++;
-    this[kNewListener](root.size, type, listener, once, capture, passive, weak);
+    addAnEventListener(this, {
+      capture,
+      passive,
+      once,
+      signal,
+      callback: typeof callback === 'function' ? callback : callback.handleEvent,
```

Comment:
Can you avoid calling `typeof`? string comparisons are always slow. I recommend avoiding it as much as you can in any js code.

Reply:
I guess I could do callback?.handleEvent ?? callback but I doubt it's faster(?)